src/blueprint.zig

```diff
@@ -0,0 +1,248 @@
+	blocks: Array3D(Block),
+
+	pub fn init(allocator: NeverFailingAllocator) Blueprint {
+		return .{.blocks = .init(allocator, 0, 0, 0)};
+	}
+	pub fn deinit(self: Blueprint, allocator: NeverFailingAllocator) void {
+		self.blocks.deinit(allocator);
+	}
+	pub fn clone(self: *Blueprint, allocator: NeverFailingAllocator) Blueprint {
+		return .{.blocks = self.blocks.clone(allocator)};
+	}
+	const CaptureResult = union(enum) {
+		success: Blueprint,
+		failure: struct {pos: Vec3i, message: []const u8},
+	};
+
+	pub fn capture(allocator: NeverFailingAllocator, pos1: Vec3i, pos2: Vec3i) CaptureResult {
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: u32 = @intCast(endX - startX + 1);
+		const sizeY: u32 = @intCast(endY - startY + 1);
+		const sizeZ: u32 = @intCast(endZ - startZ + 1);
+
+		const self = Blueprint{.blocks = .init(allocator, sizeX, sizeY, sizeZ)};
+
+		for(0..sizeX) |x| {
+			const worldX = startX +% @as(i32, @intCast(x));
+
+			for(0..sizeY) |y| {
+				const worldY = startY +% @as(i32, @intCast(y));
+
+				for(0..sizeZ) |z| {
+					const worldZ = startZ +% @as(i32, @intCast(z));
+
+					const maybeBlock = main.server.world.?.getBlock(worldX, worldY, worldZ);
+					if(maybeBlock) |block| {
+						self.blocks.set(x, y, z, block);
+					} else {
+						return .{.failure = .{.pos = .{worldX, worldY, worldZ}, .message = "Chunk containing block not loaded."}};
+					}
+				}
+			}
+		}
+		return .{.success = self};
+	}
+	pub fn paste(self: Blueprint, pos: Vec3i) void {
+		const startX = pos[0];
+		const startY = pos[1];
+		const startZ = pos[2];
+
+		for(0..self.blocks.width) |x| {
+			const worldX = startX +% @as(i32, @intCast(x));
+
+			for(0..self.blocks.depth) |y| {
+				const worldY = startY +% @as(i32, @intCast(y));
+
+				for(0..self.blocks.height) |z| {
+					const worldZ = startZ +% @as(i32, @intCast(z));
+
+					const block = self.blocks.get(x, y, z);
+					_ = main.server.world.?.updateBlock(worldX, worldY, worldZ, block);
+				}
+			}
+		}
+	}
+	pub fn load(allocator: NeverFailingAllocator, inputBuffer: []u8) !Blueprint {
+		var compressedReader = BinaryReader.init(inputBuffer, .big);
+		const version = try compressedReader.readInt(u16);
+
+		if(version > blueprintVersion) {
+			std.log.err("Blueprint version {d} is not supported. Current version is {d}.", .{version, blueprintVersion});
+			return error.UnsupportedVersion;
+		}
+		const compression = try compressedReader.readEnum(BlueprintCompression);
+		const blockPaletteSizeBytes = try compressedReader.readInt(u32);
+		const paletteBlockCount = try compressedReader.readInt(u16);
+		const width: u32 = try compressedReader.readInt(u16);
+		const depth: u32 = try compressedReader.readInt(u16);
+		const height: u32 = try compressedReader.readInt(u16);
```

Comment:
The explicit widening of the type doesn't make any here in my opinion.
```suggestion
		const width = try compressedReader.readInt(u16);
		const depth = try compressedReader.readInt(u16);
		const height = try compressedReader.readInt(u16);
```